mypyc/genops.py

```diff
@@ -382,26 +382,56 @@ def __init__(self, fitem: FuncItem = INVALID_FUNC_DEF, name: str = '',
         self.fitem = fitem
         self.name = name
         self.ns = namespace
-        # Callable classes are ClassIR instances implementing the '__call__' method, used to
-        # represent functions that are nested inside of other functions.
-        self.callable_class = INVALID_CLASS
+        # Callable classes implement the '__call__' method, and are used to represent functions
+        # that are nested inside of other functions.
+        self.callable_class = CallableClass()
         # Environment classes are ClassIR instances that contain attributes representing the
         # variables in the environment of the function they correspond to. Environment classes are
         # generated for functions that contain nested functions.
         self.env_class = INVALID_CLASS
-        # The register associated with the 'self' instance for function classes.
-        self.self_reg = INVALID_VALUE  # type: Value
-        # Environment class registers are the local registers associated with instances of an
-        # environment class, used for getting and setting attributes. env_reg is the register
-        # associated with the current environment, and prev_env_reg is the self.__mypyc_env__ field
-        # associated with the previous environment.
-        self.env_reg = INVALID_VALUE  # type: Value
-        self.prev_env_reg = INVALID_VALUE  # type: Value
         # These are flags denoting whether a given function is nested or contains a nested
         # function.
+        # Generator classes implement the '__next__' method, and are used to represent generators
+        # returned by generator functions.
+        self.generator_class = GeneratorClass()
+        # These are flags denoting whether a given function is nested, contains a nested function,
+        # or is a generator function.
         self.is_nested = False
         self.contains_nested = False
+        self.is_generator = fitem.is_generator
         # TODO: add field for ret_type: RType = none_rprimitive
+        self.curr_env_reg = INVALID_VALUE  # type: Value
+
+
+class ClassInfo(object):
```

Comment:
Find a better name for this, and document it a bit. The name `ClassInfo` makes it sound like it should exist for every class, when it is actually specifically for bookkeeping about classes that we generate that don't correspond to real source-level classes. That is, every thing that is a class in the source program *won't* have a `ClassInfo` that corresponds to it.